daft/catalog/__glue.py

```diff
@@ -0,0 +1,586 @@
+        """Lists namespaces (databases) in AWS Glue.
+
+        Args:
+            pattern (str | None): Pattern is NOT supported by Glue.
+
+        Returns:
+            list[Identifier]: List of namespace identifiers.
+        """
+        if pattern is not None:
+            # Glue may add some kind of pattern to get_databases, then we can use their scheme.
+            # Rather than making something up now which could later be broken.
+            raise ValueError("GlueCatalog does not support using pattern filters with get_databases.")
+
+        try:
+            req = {}  # type: ignore
+            namespaces = []
+            while True:
+                res = self._client.get_databases(**req)
+                for database in res["DatabaseList"]:
+                    namespaces.append(Identifier(database["Name"]))
+                if next_token := res.get("NextToken"):
+                    req["NextToken"] = next_token
+                else:
+                    break
+            return namespaces
+        except ClientError as e:
+            raise ValueError("Failed to list namespaces.") from e
+
+    def list_tables(self, pattern: str | None = None) -> list[str]:
+        if pattern is None:
+            raise ValueError("GlueCatalog requires the pattern to contain a namespace.")
+
+        req = {}
+        if "." in pattern:
+            database_name, expression = pattern.split(".", 1)
+            req["DatabaseName"] = database_name
+            req["Expression"] = expression
+        else:
+            req["DatabaseName"] = pattern
+
+        try:
+            tables = []
+            while True:
+                res = self._client.get_tables(**req)
+                for table in res["TableList"]:
+                    tables.append(f'{table["DatabaseName"]}.{table["Name"]}')
+                if next_token := res.get("NextToken"):
+                    req["NextToken"] = next_token
+                else:
+                    break
+            return tables
+        except ClientError as e:
+            raise ValueError("Failed to list tables.") from e
+
+
+class GlueTable(Table, ABC):
+    """GlueTable is the base class for various Glue format."""
+
+    _catalog: GlueCatalog
+    _table: GlueTableInfo
+
+    def __init__(self, catalog: GlueCatalog, table: GlueTableInfo):
+        self._catalog = catalog
+        self._table = table
+
+    @classmethod
+    @abstractmethod
+    def from_table_info(cls, catalog: GlueCatalog, table: GlueTableInfo) -> GlueTable:
+        """Returns a GlueTable if the GlueTableInfo matches this implementation's requirements, otherwise raise a ValueError."""
+
+    @property
+    def name(self) -> str:
+        return self._table["Name"]
+
+    def __repr__(self) -> str:
+        import json
+
+        return json.dumps(self._table, indent=4, default=str)
+
+
+class GlueCsvTable(GlueTable):
+    """GlueTable implemented by s."""
```

Comment:
What does `implemented by s` refer to?

Reply:
Originally read "by some csv files under a common s3 prefix" or something similar. I messed up docs once I started implementing more. I'll fix all this before merging!